pkg/specs/upstream.go

```diff
@@ -0,0 +1,76 @@
+package specs
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/hashicorp/go-version"
+	"github.com/pkg/errors"
+	"github.com/replicatedhq/ship/pkg/state"
+	"github.com/replicatedhq/ship/pkg/util"
+)
+
+const (
+	UpstreamVersionToken = "_latest_"
+)
+
+func (r *Resolver) MaybeResolveVersionedUpstream(ctx context.Context, upstream string, existingState state.State) (string, error) {
+	debug := level.Debug(log.With(r.Logger, "method", "resolveVersionedUpstream"))
+
+	maybeVersionedUpstream := upstream
+
+	if util.IsGithubURL(upstream) {
+		debug.Log("event", "resolve latest release")
+		latestReleaseVersion, err := r.GitHubFetcher.ResolveLatestRelease(ctx, upstream)
+		if err != nil {
+			if strings.Contains(upstream, UpstreamVersionToken) {
+				return "", errors.Wrap(err, "resolve latest release")
+			}
+			return upstream, nil
+		}
+
+		maybeVersionedUpstream = strings.Replace(upstream, UpstreamVersionToken, latestReleaseVersion, 1)
+
+		debug.Log("event", "check previous version")
+		if existingState.Versioned().V1.Metadata != nil && existingState.Versioned().V1.Metadata.Version != "" {
+			if strings.Contains(upstream, UpstreamVersionToken) {
+				previousVersion, err := version.NewVersion(existingState.Versioned().V1.Metadata.Version)
+				if err != nil {
+					return maybeVersionedUpstream, nil
+				}
+
+				latestVersion, err := version.NewVersion(latestReleaseVersion)
```

Comment:
`latestVersion` is never returned

Reply:
no the intent is to return `maybeVersionedUpstream`, `latestVersion` is only a `go-version` version used for comparison purposes